src/models/auth.js

```diff
@@ -15,36 +11,34 @@ const POLICIES = {
   SEND_MESSAGES: 'send_messages',
 };
 
-function hasPermission(user, p) {
-  // Chris github - BirdTho 9/21 user may fail if called when logged out, as in the case in Context.js
-  // console.assert(user, "Why user fail?", user);
-  if (!user) return false;
-  if (p instanceof Array) {
-    return p.some((permission) => {
-      return user.policy.policies.some((r) => r.name === permission);
-    });
-  } else {
-    return user.policy.policies.some((r) => r.name === p) ? true : false;
-  }
+function hasPermission(permissions) {
+  const {
+    keycloak: { tokenParsed },
+  } = useKeycloak();
+  const roles = tokenParsed?.realm_access?.roles;
+  return permissions
+    .map((permission) => roles?.includes(permission))
+    .reduce((a, b) => a && b);
```

Comment:
@TANguyen1893 I think it might be limited to just using role to check the permission? Seems we need a permission-role-policy pattern to define a fine-grained system? like:
policy A defined that role B has the permission to doing action C
like: 
The tree-user has permission to view the list of tree
The tree-manager has the permission to edit the list of tree

Reply:
@dadiorchen You are right. It would be much better to have a fine-grained permission system. I would prefer that than just simply roles.